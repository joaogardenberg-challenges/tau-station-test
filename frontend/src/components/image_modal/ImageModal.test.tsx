import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import range from 'lodash/range'
import mapKeys from 'lodash/mapKeys'
import ThemeProvider from 'providers/Theme'

const useHistory = jest.fn()
const useParams = jest.fn()
const useSelector = jest.fn()
const useQuery = jest.fn()
const createPortal = (children: any) => children
const CSSTransition = ({ children }: { children: any }) => children

jest.doMock('react-router-dom', () => ({ useHistory, useParams }))
jest.doMock('react-redux', () => ({ useSelector }))
jest.doMock('hooks/useQuery', () => useQuery)
jest.doMock('react-dom', () => ({ createPortal }))

jest.doMock('react-transition-group', () => ({
  ...jest.requireActual('react-transition-group'),
  CSSTransition
}))

jest.mock('./Image')
jest.mock('./ImageSizes')
jest.mock('./ImagesPreloader')
jest.mock('./ImagesFetcher')

const ImageModal = require('./ImageModal').default

const renderWithTheme = (ui: any, options: any = {}) =>
  render(ui, { wrapper: ThemeProvider, ...options })

describe('Image Modal', () => {
  const queryToString = 'query'
  const defaultSelectedId = 2

  const defaultSelector = {
    imagesList: mapKeys(
      range(defaultSelectedId - 1, defaultSelectedId + 2).map((id) => ({ id })),
      'id'
    ),
    isFetching: false
  }

  beforeEach(() => {
    useHistory.mockReturnValue({ push: jest.fn() })
    useParams.mockReturnValue({ id: defaultSelectedId })
    useSelector.mockReturnValue(defaultSelector)

    useQuery.mockReturnValue({
      get: () => '400',
      toString: () => queryToString
    })
  })

  it('renders successfully', () => {
    const { container } = renderWithTheme(<ImageModal />)

    const image = screen.getByTestId('image')
    const imageSizes = screen.getByTestId('image-sizes')
    const imagesPreloader = screen.getByTestId('images-preloader')
    const imagesFetcher = screen.getByTestId('images-fetcher')

    expect(image).toBeInTheDocument()
    expect(imageSizes).toBeInTheDocument()
    expect(imagesPreloader).toBeInTheDocument()
    expect(imagesFetcher).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('closes the modal', () => {
    renderWithTheme(<ImageModal />)

    const closeButton = screen.getByRole('button', { name: 'close' })
    userEvent.click(closeButton)

    expect(useHistory().push).toHaveBeenCalledTimes(1)
    expect(useHistory().push).toHaveBeenCalledWith('/images')
  })

  it('has the correct previous button disabled state', () => {
    const { rerender } = renderWithTheme(<ImageModal />)

    const prevButton = screen.getByRole('button', { name: 'previous' })
    expect(prevButton).not.toHaveAttribute('disabled')

    useParams.mockReturnValue({ id: defaultSelectedId - 1 })
    rerender(<ImageModal />)
    expect(prevButton).toHaveAttribute('disabled')
  })

  it('has the correct next button disabled state', () => {
    const { rerender } = renderWithTheme(<ImageModal />)

    const nextButton = screen.getByRole('button', { name: 'next' })
    expect(nextButton).not.toHaveAttribute('disabled')

    useParams.mockReturnValue({ id: defaultSelectedId + 1 })
    rerender(<ImageModal />)
    expect(nextButton).toHaveAttribute('disabled')
  })

  it('navigates to the previous image', () => {
    renderWithTheme(<ImageModal />)

    const prevButton = screen.getByRole('button', { name: 'previous' })
    userEvent.click(prevButton)

    expect(useHistory().push).toHaveBeenCalledTimes(1)
    expect(useHistory().push).toHaveBeenCalledWith({
      pathname: `/images/${defaultSelectedId - 1}`,
      search: queryToString
    })
  })

  it('navigates to the next image', () => {
    renderWithTheme(<ImageModal />)

    const nextButton = screen.getByRole('button', { name: 'next' })
    userEvent.click(nextButton)

    expect(useHistory().push).toHaveBeenCalledTimes(1)
    expect(useHistory().push).toHaveBeenCalledWith({
      pathname: `/images/${defaultSelectedId + 1}`,
      search: queryToString
    })
  })
})
