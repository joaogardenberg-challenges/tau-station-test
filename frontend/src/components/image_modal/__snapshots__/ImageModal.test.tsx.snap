// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Modal renders successfully 1`] = `
<div>
  <aside
    class="sc-bdfBQB ekhtQf"
  >
    <div
      class="carousel"
    >
      <div
        class="image"
      />
      <button
        aria-label="next"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium sc-gsTEea jekxUr css-1ns4028-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="previous"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium sc-gsTEea jekxUr css-1ns4028-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <button
      aria-label="close"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge sc-gsTEea sc-dlfnuX jekxUr bknlWH css-1q5do0g-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="image-sizes"
    />
    <div
      class="images-preloader"
    />
    <div
      class="images-fetcher"
    />
  </aside>
</div>
`;
